packages/react-dom/src/__tests__/ReactBrowserEventEmitter-test.internal.js

```diff
@@ -9,360 +9,480 @@
 
-    function renderTree() {
-      ReactDOM.render(
-        <div ref={c => (GRANDPARENT = c)} {...GRANDPARENT_PROPS}>
-          <div ref={c => (PARENT = c)} {...PARENT_PROPS}>
-            <ChildWrapper {...CHILD_PROPS} />
-          </div>
-        </div>,
+    createInitialInstance = () => {
+      const initialInstance = ReactDOM.render(
+        <Parent parentProps={{}} childProps={{onClick: LISTENER}} />,
         container,
       );
-    }
+      expect(childRef.props.onClick).toBe(LISTENER);
 
-    renderTree();
-
-    getListener = function(node, eventName) {
-      const inst = ReactDOMComponentTree.getInstanceFromNode(node);
-      return EventPluginHub.getListener(inst, eventName);
-    };
-    putListener = function(node, eventName, listener) {
-      switch (node) {
-        case CHILD:
-          CHILD_PROPS[eventName] = listener;
-          break;
-        case PARENT:
-          PARENT_PROPS[eventName] = listener;
-          break;
-        case GRANDPARENT:
-          GRANDPARENT_PROPS[eventName] = listener;
-          break;
-      }
-      // Rerender with new event listeners
-      renderTree();
-    };
-    deleteAllListeners = function(node) {
-      switch (node) {
-        case CHILD:
-          CHILD_PROPS = {};
-          break;
-        case PARENT:
-          PARENT_PROPS = {};
-          break;
-        case GRANDPARENT:
-          GRANDPARENT_PROPS = {};
-          break;
-      }
-      renderTree();
+      return initialInstance;
     };
 
-    idCallOrder = [];
+    updateInstance = ({parentProps, childProps}) => {
+      return ReactDOM.render(
+        <Parent parentProps={parentProps} childProps={childProps} />,
+        container,
+      );
+    };
   });
 
   afterEach(() => {
     document.body.removeChild(container);
     container = null;
+    childRef = null;
   });
 
   it('should store a listener correctly', () => {
-    registerSimpleTestHandler();
-    const listener = getListener(CHILD, ON_CLICK_KEY);
-    expect(listener).toBe(LISTENER);
+    createInitialInstance();
+    updateInstance({
+      parentProps: {},
+      childProps: {
+        onClick: LISTENER,
+      },
+    });
+    expect(childRef.props.onClick).toBe(LISTENER);
   });
 
   it('should retrieve a listener correctly', () => {
-    registerSimpleTestHandler();
-    const listener = getListener(CHILD, ON_CLICK_KEY);
-    expect(listener).toEqual(LISTENER);
+    createInitialInstance();
+    updateInstance({
+      parentProps: {},
+      childProps: {
+        onClick: LISTENER,
+      },
+    });
+    expect(childRef.props.onClick).toBe(LISTENER);
   });
 
   it('should clear all handlers when asked to', () => {
-    registerSimpleTestHandler();
-    deleteAllListeners(CHILD);
-    const listener = getListener(CHILD, ON_CLICK_KEY);
-    expect(listener).toBe(undefined);
+    createInitialInstance();
+    updateInstance({
+      parentProps: {},
+      childProps: {},
+    });
+    expect(childRef.props.onClick).toBe(undefined);
```

Comment:
This test is again only asserting the props that you set. I think you can use ReactDOM.render() with an `onClick` listener and then another ReactDOM.render() without the listener, then fire a native event, and assert that the handler is not called. This way we know that React has cleaned up the listeners internally.